file.h

```diff
@@ -1455,6 +1367,122 @@ std::uint64_t basic_file_base<File>::tell() const
+    data.resize(actually_read);
+
+    // Consider omitting
+    data.shrink_to_fit();
+
+    return data;
+}
+
+template <typename File>
+std::vector<std::byte> basic_file_base<File>::read_zero() const
+{
+    // 0 means that the caller is requesting to read all available
+    // try to find out how many bytes can be read
+
+    std::optional<std::size_t> bytes_can_be_read = try_get_bytes_can_be_read();
+
+    if (!bytes_can_be_read.has_value()) {
+        // Cannot determine apriori how many bytes can be read.
+        return read_and_resize_to_end();
+    }
+
+    auto bytes_to_read = bytes_can_be_read.value();
+    if (bytes_to_read == 0) {
+        // Nothing to read.
+        return {};
+    }
+    return read_nonzero(bytes_to_read);
+}
+
+template <typename File>
+std::optional<std::size_t> basic_file_base<File>::try_get_bytes_can_be_read() const
+{
+    auto file_size = this->size();
+    if (file_size == 0) {
+        // Cannot determine file size
+        return std::nullopt;
+    }
+    auto current_offset = tell();
+    if (tell() >= file_size) {
+        // At or beyond the end. No bytes can be read.
+        return 0;
+    }
+    auto bytes_can_be_read = file_size - current_offset;
+
+    if (bytes_can_be_read > (std::numeric_limits<std::size_t>::max)()) {
+        throw std::range_error(
+            "Amount of file data is too large for this platform.");
+    }
+
+    // This positive number of bytes are available for read on this platform
+    return bytes_can_be_read;
+}
+
+template <typename File>
+std::size_t basic_file_base<File>::get_bigger_size(std::size_t curr_size) const
+{
+    constexpr std::size_t initial_vector_size = 0x1000;
+
+    if (curr_size < 4) { 
+        // assure that the value returned is bigger than curr_size
+        return initial_vector_size;
+    }
+
+    constexpr auto resize_factor_limit =
+        (std::numeric_limits<std::size_t>::max)() / 3 * 2;
+
+    if (curr_size >= resize_factor_limit) {
+        return (std::numeric_limits<std::size_t>::max)();
+    }
+
+    return curr_size / 2 * 3;
+}
+
+template <typename File>
+std::vector<std::byte> basic_file_base<File>::read_and_resize_to_end() const
+{
+    std::vector<std::byte> data;
+    std::size_t total_read = 0;
+
+    // As we don't know how many bytes are available for read, loop
+    // while reading and extending the vector.
+
+    while (true) {
+        data.resize(get_bigger_size(data.size()));
+
+        std::size_t to_read_now = data.size() - total_read;
+
+        auto read_now = read({ data.data() + total_read, to_read_now });
+
+        total_read += read_now;
+
+        if (read_now < to_read_now) {
+            // Reached the end
+            break;
+        }
+
+        if (data.size() == (std::numeric_limits<std::size_t>::max)()) {
+            // Differentiate between reached the end and not.
+            std::byte one_byte;
+            if (1 == read({ &one_byte, 1 })) {
```

Comment:
I would really like to avoid this other read call.

Reply:
OK.  But a comment might be in order.